# Exercise 02 -  OLAP Cubes - Roll Up and Drill Down

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [2]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

 set_config 
------------
 
(1 row)

 setval 
--------
    200
(1 row)

 setval 
--------
    605
(1 row)

 setval 
--------
     16
(1 row)

 setval 
--------
    600
(1 row)

 setval 
--------
    109
(1 row)

 setval 
--------
    599
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
   1000
(1 row)

 setval 
--------
   4581
(1 row)

 setval 
--------
      6
(1 row)

 setval 
--------
  32098
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
      2
(1 row)

 setval 
--------
      2
(1 row)



### Connect to the local database where Pagila is loaded

In [3]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

## Roll-up
- Stepping up the level of aggregation to a large grouping
- e.g.`city` is summed as `country`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and country. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [4]:
%%time
%%sql

SELECT dimdate.day as day, dimmovie.rating as rating, dimcustomer.country as country, sum(sales_amount) as revenue 
FROM factsales
JOIN dimdate ON (factsales.date_key = dimdate.date_key) 
JOIN dimmovie ON (factsales.movie_key = dimmovie.movie_key)
JOIN dimcustomer ON (factsales.customer_key = dimcustomer.customer_key)
WHERE rating = 'PG-13'
group by (day, rating, country)
ORDER BY revenue desc
LIMIT 20;


 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 5.29 ms, sys: 0 ns, total: 5.29 ms
Wall time: 16.2 ms


[(30, 'PG-13', 'China', Decimal('146.67')),
 (21, 'PG-13', 'India', Decimal('128.74')),
 (18, 'PG-13', 'India', Decimal('121.72')),
 (30, 'PG-13', 'India', Decimal('116.72')),
 (17, 'PG-13', 'China', Decimal('111.75')),
 (20, 'PG-13', 'India', Decimal('101.79')),
 (27, 'PG-13', 'India', Decimal('98.79')),
 (18, 'PG-13', 'Japan', Decimal('96.79')),
 (30, 'PG-13', 'Russian Federation', Decimal('94.80')),
 (18, 'PG-13', 'China', Decimal('94.78')),
 (30, 'PG-13', 'Philippines', Decimal('86.81')),
 (30, 'PG-13', 'Mexico', Decimal('83.80')),
 (28, 'PG-13', 'China', Decimal('81.81')),
 (19, 'PG-13', 'China', Decimal('80.82')),
 (29, 'PG-13', 'India', Decimal('79.83')),
 (17, 'PG-13', 'United States', Decimal('77.83')),
 (19, 'PG-13', 'India', Decimal('77.80')),
 (20, 'PG-13', 'Philippines', Decimal('73.85')),
 (30, 'PG-13', 'United States', Decimal('72.80')),
 (28, 'PG-13', 'Japan', Decimal('67.86'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>China</td>
        <td>169.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>India</td>
        <td>156.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>India</td>
        <td>153.64</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>China</td>
        <td>146.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>R</td>
        <td>China</td>
        <td>145.66</td>
    </tr>
</tbody></table></div>

## Drill-down
- Breaking up one of the dimensions to a lower level.
- e.g.`city` is broken up into  `districts`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and district. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [5]:
%%time
%%sql

SELECT dimdate.day as day, dimmovie.rating as rating, dimcustomer.district as district, sum(sales_amount) as revenue 
FROM factsales
JOIN dimdate ON (factsales.date_key = dimdate.date_key) 
JOIN dimmovie ON (factsales.movie_key = dimmovie.movie_key)
JOIN dimcustomer ON (factsales.customer_key = dimcustomer.customer_key)
WHERE rating = 'PG-13'
group by (day, rating, district)
ORDER BY revenue desc
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.76 ms, sys: 0 ns, total: 4.76 ms
Wall time: 14.8 ms


[(30, 'PG-13', 'Southern Tagalog', Decimal('53.88')),
 (17, 'PG-13', 'Shandong', Decimal('34.95')),
 (21, 'PG-13', 'Uttar Pradesh', Decimal('29.94')),
 (21, 'PG-13', 'Southern Tagalog', Decimal('28.96')),
 (20, 'PG-13', 'Buenos Aires', Decimal('28.95')),
 (27, 'PG-13', 'Buenos Aires', Decimal('28.92')),
 (30, 'PG-13', 'Uttar Pradesh', Decimal('27.94')),
 (2, 'PG-13', 'So Paulo', Decimal('26.96')),
 (30, 'PG-13', 'Buenos Aires', Decimal('25.94')),
 (28, 'PG-13', 'England', Decimal('24.96')),
 (28, 'PG-13', 'Dhaka', Decimal('24.96')),
 (20, 'PG-13', 'Southern Tagalog', Decimal('23.97')),
 (30, 'PG-13', 'Karnataka', Decimal('22.96')),
 (20, 'PG-13', 'West Java', Decimal('22.95')),
 (17, 'PG-13', 'Ondo & Ekiti', Decimal('22.94')),
 (30, 'PG-13', 'Zanzibar West', Decimal('21.97')),
 (21, 'PG-13', 'Maharashtra', Decimal('21.96')),
 (19, 'PG-13', 'Central Java', Decimal('21.96')),
 (23, 'PG-13', 'Liaoning', Decimal('21.95')),
 (17, 'PG-13', 'Southern Tagalog', Decimal('21.95'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>district</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Southern Tagalog</td>
        <td>53.88</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Inner Mongolia</td>
        <td>38.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Shandong</td>
        <td>36.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>West Bengali</td>
        <td>36.92</td>
    </tr>
    <tr>
        <td>17</td>
        <td>PG-13</td>
        <td>Shandong</td>
        <td>34.95</td>
    </tr>
</tbody></table></div>